core/solver/batch_bicgstab_kernels.hpp

```diff
@@ -90,6 +90,112 @@ inline int local_memory_requirement(const int num_rows, const int num_rhs)
 }
 
 
+struct storage_config {
+    // preconditioner storage
+    bool prec_shared;
+    // total number of shared vectors
+    int n_shared;
+    // number of vectors in global memory
+    int n_global;
+    // global stride from one batch entry to the next
+    int gmem_stride_bytes;
+    // padded vector length
+    int padded_vec_len;
+};
+
+
+template <int align_bytes>
+void set_gmem_stride_bytes(storage_config& sconf,
+                           const int multi_vector_size_bytes,
+                           const int prec_storage_bytes)
+{
+    int gmem_stride = sconf.n_global * multi_vector_size_bytes;
+    if (!sconf.prec_shared) {
+        gmem_stride += prec_storage_bytes;
+    }
+    // align global memory chunks
+    sconf.gmem_stride_bytes =
+        gmem_stride > 0 ? ((gmem_stride - 1) / align_bytes + 1) * align_bytes
+                        : 0;
+}
+
+
+/**
+ * Calculates the amount of in-solver storage needed by batch-Bicgstab and
+ * the split between shared and global memory.
+ *
+ * The calculation includes multivectors for
+ * - r
+ * - r_hat
+ * - p
+ * - p_hat
+ * - v
+ * - s
+ * - s_hat
+ * - t
+ * - x
+ * In addition, small arrays are needed for
+ * - rho_old
+ * - rho_new
+ * - omega
+ * - alpha
+ * - temp
+ * - rhs_norms
+ * - res_norms
+ *
+ * @param shared_mem_per_blk  The amount of shared memory per block to use for
+ *   keeping intermediate vectors. In case keeping the matrix in L1 cache etc.
+ *   should be prioritized, the cache configuration must be updated separately
+ *   and the needed space should be subtracted before passing to this
+ *   function.
+ * @param num_rows  Size of the matrix.
+ * @param num_nz  Number of nonzeros in the matrix
+ * @param num_rhs  Number of right-hand-sides in the vectors.
+ * @return  A struct containing allocation information specific to Bicgstab.
+ */
+template <typename Prectype, typename ValueType, int align_bytes = 32>
+storage_config compute_shared_storage(const int shared_mem_per_blk,
+                                      const int num_rows, const int num_nz,
+                                      const int num_rhs)
+{
+    using real_type = remove_complex<ValueType>;
+    const int vec_size = num_rows * num_rhs * sizeof(ValueType);
+    const int num_main_vecs = 9;
+    const int prec_storage =
+        Prectype::dynamic_work_size(num_rows, num_nz) * sizeof(ValueType);
+    int rem_shared = shared_mem_per_blk;
+    // Set default values. All vecs are in global.
+    storage_config sconf{false, 0, num_main_vecs, 0, num_rows};
+    // If available shared mem, is zero, set all vecs to global.
```

Comment:
That comment doesn't describe what the code below does.

Reply:
If the remaining shared is zero, then in the function, we have `sconf.n_global` set to all vectors, and we allocate everything on global memory. Maybe 
```
// If available shared memory is zero, then set all vectors to be in global memory, including the preconditioner workspace 
```
is more clear ?

Reply:
This is again an issue with the early return. None of the code in the if-branch sets `sconf.n_global`. It just uses the default setting from above. Again, it would be much easier to describe what is happening in this function without the early return.

Reply:
The function could be interpreted as a sequence of operations like:
```
set_number_of_vectors(...);
set_preconditioner(...);
set_global_memory_stride(...);
```
But the early return makes that hard to follow (internally these functions could use early return).
Maybe this can help to illustrate my difficulty understanding the control flow. If you put in the corresponding else-branch for each if-branch, then the final return is *three* levels deep. IMO, something like this is hard to follow.

Reply:
IMO this type of early return makes the code easier to read, but I understand your concern. I guess it should be possible to refactor this into separate functions like you suggested such that the `set_num_vec...`, `set_prec...` take `rem_shared` and `sconf` and set the values and return `sconf`, but maybe we postpone this refactoring for now. 

But I think eventually I would like to unify this across all batched solver classes, because essentially the computation is similar. I have noted it down in #1376 and we can come back to this later.